File: leetcode/ag_282.py
class Solution(object):

    def addOperators(self, num, target):
        """
        :type num: str
        :type target: int
        :rtype: List[str]
        """
        ans = []
        self.solve( num  , target , ans , 0 ,"" ,0 ,0)
        return ans

    def solve(self , num , target , retList , pos , path , current , prev ):
        if len(num) == pos:
            if current == target:
                retList.append(path)
                return
        
        for i in range(pos , len(num)):
            if i != pos and num[pos] == '0':
                break;
            curstr = num[pos:i+1]
            currentval = int(curstr)
            if pos == 0:
                self.solve(num , target , retList , i+1 , path + curstr , current + currentval , currentval)
            else:
                self.solve(num , target , retList , i+1 , path + '+' + curstr , current + currentval , currentval)
                self.solve(num , target , retList , i+1 , path + '-' + curstr , current - currentval , -currentval)
                self.solve(num , target , retList , i+1 , path + '*' + curstr , current - prev + prev * currentval , prev * currentval)

if __name__=='__main__':
    sn = Solution()
    num = "105"
    target = 5
    print sn.addOperators(num , target)

File: leetcode/ag_282.cpp
#include <iostream>
#include <vector>


using namespace std;


class Solution {
public:
    vector<string> addOperators(string num, int target) {
        vector<string> sol;
        if (num.empty()) {
            return sol;
        }
        helper(sol, "", num, target, 0, 0, 0);
        return sol;
    }
    void helper(vector<string>& sol, string path, string num, int target, int pos, long eval, long multed) {
        if (pos == num.length()) {
            if (target == eval) {
                sol.push_back(path);
            }
            return;
        }
        for (int i = pos; i < num.length(); i++) {
            if (i != pos && num[pos] == '0') {
                break;
            }
            string curString = num.substr(pos, i - pos + 1);
            auto cur = stol(curString);
            if (pos == 0) {
                helper(sol, path + curString, num, target, i + 1, cur, cur);
            } else {
                helper(sol, path + "+" + curString, num, target, i + 1, eval + cur, cur);
                 
                helper(sol, path + "-" + curString, num, target, i + 1, eval - cur, -cur);
                 
                helper(sol, path + "*" + curString, num, target, i + 1, eval - multed + multed * cur, multed * cur);
            }
        }
    }
};

int main()
{

}
